java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -115,14 +120,37 @@ public BooleanConstraint shouldNotInverse() {
 
   @Override
   public List<ProgramState> setConstraint(ProgramState initialProgramState, BooleanConstraint booleanConstraint) {
-    ProgramState programState = checkRelation(booleanConstraint, initialProgramState);
-    if (programState == null) {
+    if (!checkRelation(booleanConstraint, initialProgramState)) {
       return ImmutableList.of();
     }
+    return booleanConstraint == BooleanConstraint.TRUE ? getNewProgramStates(initialProgramState) : inverse().getNewProgramStates(initialProgramState);
+  }
+
+  private List<ProgramState> getNewProgramStates(ProgramState initialProgramState) {
+    List<SymbolicValue> newRelations = transitiveRelations(initialProgramState);
+    newRelations.add(this);
+
+    List<ProgramState> programStates = new ArrayList<>();
+    programStates.add(initialProgramState);
+    for (SymbolicValue relationalSymbolicValue : newRelations) {
+      List<ProgramState> intermediateStates = new ArrayList<>();
+      for (ProgramState programState: programStates) {
+        intermediateStates.addAll(relationalSymbolicValue.copyAllConstraints(BooleanConstraint.TRUE, programState));
+      }
+      programStates = intermediateStates;
+    }
+    return programStates;
+  }
+
+  protected RelationalSymbolicValue inverse() {
+    return new RelationalSymbolicValue(kind.inverse(), leftOp, rightOp);
+  }
+
+  protected List<ProgramState> copyAllConstraints(BooleanConstraint booleanConstraint, ProgramState programState) {
```

Comment:
I tend to think that this method should not anymore get the boolean constraint as a parameter. Quite some flaws raised by SonarTech can be solved by removing this param.